postprocessing/pcap2mjr.1
.TH PCAP2MJR 1
.SH NAME
pcap2mjr \- Helper tool to convert a pcap dump to a Janus recording.
.SH SYNOPSIS
.B pcap2mjr [options]
.IR source.pcap
.IR destination.mjr
.SH DESCRIPTION
.B pcap2mjr
is a simple utility that allows you take .pcap network captures, extract a specific RTP session via its SSRC, and convert it to an .mjr Janus recording instead. Its main purpose is helping convert .pcap captures to media files, or make it easier to replay them via Janus.
.TP
The tool requires a path to the .pcap file to read, and a path to the target .pcap file; besides, it needs info on the codec and the SSRC to filter. Notice that if the tool can't detect any RTP packet with that SSRC, it will result in an error.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
Print help and exit
.TP
.BR \-V ", " \-\-version
Print version and exit
.TP
.BR \-c ", " \-\-codec=\fIcodec\fR
Codec the recording will contain (e.g., opus, vp8, etc.)
.TP
.BR \-s ", " \-\-ssrc=\fISSRC (numeric)\fR
SSRC of the packets in the pcap file to save
.TP
.BR \-w ", " \-\-warnings
Show warnings for skipped packets (e.g., not RTP or wrong SSRC)
.SH EXAMPLES
\fBpcap2mjr -c opus -s 12345678 rec1234.pcap rec1234.mjr\fR \- Read all RTP packets with SSRC 12345678 from the provided .pcap file, and save them to a new .mjr file as an Opus recording
.SH BUGS
.TP
If you think you found a bug or want to contribute a feature, you can issue or a pull request on https://github.com/meetecho/janus-gateway/issues.
.TP
Anyway, before doing that make sure you read the documentation at https://janus.conf.meetecho.com/docs/ and that it has not been discussed already at https://groups.google.com/forum/#!forum/meetecho-janus. We only use Github for code issues, and \fBNOT\fR for configuration or usage issues: use the group for that.
.SH SEE ALSO
.TP
https://github.com/meetecho/janus-gateway \- Official repository
.TP
https://janus.conf.meetecho.com \- Demos and documentation
.TP
https://groups.google.com/forum/#!forum/meetecho-janus \- Community
.TP
https://www.meetecho.com/blog/ \- Tutorials and blog posts on Janus
.SH AUTHORS
Lorenzo Miniero (lorenzo@meetecho.com)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.


## [v0.11.5] - 2021-10-18

- Add API to optionally force Janus to use TURN [[PR-2774](https://github.com/meetecho/janus-gateway/pull/2774)]
- Fixed slow path on SDP parsing [[PR-2776](https://github.com/meetecho/janus-gateway/pull/2776)]
- Added event handlers option to send stats for a PeerConnection in a single event, rather than per-media (thanks @JanFellner!) [[PR-2785](https://github.com/meetecho/janus-gateway/pull/2785)]
- Fixed occasional deadlocks on malformed requests in VideoRoom [[Issue-2780](https://github.com/meetecho/janus-gateway/issues/2780)]
- Fixed AudioBridge plain RTP thread sometimes exiting prematurely
- Fixed broken upsampling when using G.711 in AudioBridge
- Add pause/resume recording functionality to Record&Play and SIP plugins (thanks @isnumanagic!) [[PR-2724](https://github.com/meetecho/janus-gateway/pull/2724)]
- Fixed broken support for Unix Sockets in WebSockets Admin API (thanks @thatsmydoing!) [[PR-2787](https://github.com/meetecho/janus-gateway/pull/2787)]
- Added timing info for video rotation when post-processing recordings
- Added linter checks to janus.js (thanks @davel!) [[PR-2272](https://github.com/meetecho/janus-gateway/pull/2272)]
- Other smaller fixes and improvements (thanks to all who contributed pull requests and reported issues!)

## [v0.11.4] - 2021-09-06

- Fixed ICE restart issues with recent versions of libnice [[PR-2729](https://github.com/meetecho/janus-gateway/pull/2729)]
- Changed randon number generators to use crypto-safe functions (thanks @jmfotokite!) [[PR-2738](https://github.com/meetecho/janus-gateway/pull/2738)]
- Added support for abs-send-time RTP extension [[PR-2721](https://github.com/meetecho/janus-gateway/pull/2721)]
- Added configurable mechanism for manually setting static event loop to use for new handles [[PR-2684](https://github.com/meetecho/janus-gateway/pull/2684)]
- Fixed datachannel protocol not being sent to plugins for incoming messages [[Issue-2753](https://github.com/meetecho/janus-gateway/issues/2753)]
- Added ability to specify recordings folder in AudioBridge [[PR-2707](https://github.com/meetecho/janus-gateway/pull/2707)]
- Added support for forwarding groups in AudioBridge [[PR-2653](https://github.com/meetecho/janus-gateway/pull/2653)]
- Fixed missing Contact header in SIP plugin when using Sofia >= 1.13 [[PR-2708](https://github.com/meetecho/janus-gateway/pull/2708)]
- Better SDES-SRTP negotiation in SIP and NoSIP plugins [[PR-2727](https://github.com/meetecho/janus-gateway/pull/2727)]
- Fixed WebSocket transport and event handler lagging 25/30s when shutting down or reconnecting (thanks @JanFellner!) [[Issue-2734](https://github.com/meetecho/janus-gateway/issues/2734)]
- Fixed incoming_header_prefixes not working for helper sessions in SIP plugin
- Fix partial/broken ACL support in TextRoom plugin [[PR-2763](https://github.com/meetecho/janus-gateway/pull/2763)]
- Fixed potential race condition when reclaiming sessions in HTTP transport plugin
- Fixed WebSocket event handler reconnect mechanism (thanks @JanFellner!) [[PR-2736](https://github.com/meetecho/janus-gateway/pull/2736)]
- Other smaller fixes and improvements (thanks to all who contributed pull requests and reported issues!)

## [v0.11.3] - 2021-06-15

- Fixed rare crash when detaching handles [[Issue-2464](https://github.com/meetecho/janus-gateway/issues/2464)]
- Added option to offer IPv6 link-local candidates as well [[PR-2689](https://github.com/meetecho/janus-gateway/pull/2689)]
- Added spatial audio support to AudioBridge via stereo mixing [[PR-2446](https://github.com/meetecho/janus-gateway/pull/2446)]
- Added support for plain RTP participants to AudioBridge [[PR-2464](https://github.com/meetecho/janus-gateway/pull/2464)]
- Added API to start/stop AudioBridge recordings dynamically (thanks @rajneeshksoni!) [[PR-2674](https://github.com/meetecho/janus-gateway/pull/2674)]
- Fixed broken mountpoint switching when using different payload types in Streaming plugin [[PR-2692](https://github.com/meetecho/janus-gateway/pull/2692)]
- Fixed occasional deadlock on Streaming plugin mountpoint destroy during RTSP reconnects (thanks @lionelnicolas!) [[PR-2700](https://github.com/meetecho/janus-gateway/pull/2700)]
- Added "Expires" support to SUBSCRIBE in SIP plugin (thanks @nicolasduteil!) [[PR-2661](https://github.com/meetecho/janus-gateway/pull/2661)]
- Added option to specify Call-ID for SUBSCRIBE dialogs in SIP plugin (thanks @nicolasduteil!) [[PR-2664](https://github.com/meetecho/janus-gateway/pull/2664)]
- Fixed broken simulcast support in VideoCall plugin (thanks @lucily-star!) [[PR-2671](https://github.com/meetecho/janus-gateway/pull/2671)]
- Implemented RabbitMQ reconnection logic, in both transport and event handler (thanks @chriswiggins!) [[PR-2651](https://github.com/meetecho/janus-gateway/pull/2651)]
- Added support for renegotiation of external streams in janus.js (thanks @kmeyerhofer!) [[PR-2604](https://github.com/meetecho/janus-gateway/pull/2604)]
- Added support for HEVC/H.265 aggregation packets (AP) to janus-pp-rec (thanks @nu774!) [[PR-2662](https://github.com/meetecho/janus-gateway/pull/2662)]
- Refactored janus-pp-rec to cleanup the code, and use libavformat for Opus as well (thanks @lu-zero!) [[PR-2665](https://github.com/meetecho/janus-gateway/pull/2665)]
- Added additional target formats for some recorded codecs [[PR-2680](https://github.com/meetecho/janus-gateway/pull/2680)]
- Other smaller fixes and improvements (thanks to all who contributed pull requests and reported issues!)

## [v0.11.2] - 2021-05-03

- Added support for relative paths in config files, currently only in MQTT event handler (thanks @RSATom!) [[PR-2623](https://github.com/meetecho/janus-gateway/pull/2623)]
- Removed support for now deprecated frame-marking RTP extension [[PR-2640](https://github.com/meetecho/janus-gateway/pull/2640)]

postprocessing/pcap2mjr.c
/*! \file    pcap2mjr.c
 * \author   Lorenzo Miniero <lorenzo@meetecho.com>
 * \copyright GNU General Public License v3
 * \brief    Helper tool to convert .pcap files to Janus .mjr recordings
 * \details  Our Janus WebRTC gateway provides a simple helper (janus_recorder)
 * to allow plugins to record audio, video and text frames sent by users.
 * These recordings can then be processed and converted to playable files,
 * or replayed via WebRTC again. The \c pcap2mjr tool is a simple utility
 * that allows you take .pcap network captures, extract a specific RTP
 * session via its SSRC, and convert it to an .mjr Janus recording instead.
 * Its main purpose is helping convert .pcap captures to media files, or
 * make it easier to replay them via Janus.
 *
 * Using the utility is quite simple. Just pass, as arguments to the tool,
 * the SSRC to extract, the codec used for the RTP packets originally, the
 * path to the .pcap source file, and the path to the destination file, e.g.:
 *
\verbatim
./pcap2mjr -c vp8 -s 12345678 /path/to/source.pcap /path/to/destination.mjr
\endverbatim
 *
 * If the tool can't detect any RTP packet with that SSRC, it will result in an error.
 *
 * \ingroup postprocessing
 * \ref postprocessing
 */

#include <arpa/inet.h>
#ifdef __MACH__
#include <machine/endian.h>
#else
#include <endian.h>
#endif
#include <inttypes.h>
#include <string.h>
#include <stdlib.h>
#include <signal.h>
#include <sys/time.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <netinet/udp.h>
#include <errno.h>

#include <glib.h>
#include <jansson.h>

#include <pcap.h>
#include <pcap/sll.h>

#include "../debug.h"
#include "p2m-cmdline.h"
#include "pp-rtp.h"


#define htonll(x) ((1==htonl(1)) ? (x) : ((gint64)htonl((x) & 0xFFFFFFFF) << 32) | htonl((x) >> 32))
#define ntohll(x) ((1==ntohl(1)) ? (x) : ((gint64)ntohl((x) & 0xFFFFFFFF) << 32) | ntohl((x) >> 32))

int janus_log_level = 4;
gboolean janus_log_timestamps = FALSE;
gboolean janus_log_colors = TRUE;
char *janus_log_global_prefix = NULL;
int lock_debug = 0;

int working = 0;

/* Info header in the structured recording */
static const char *header = "MJR00002";
/* Frame header in the structured recording */
static const char *frame_header = "MEET";

/* Ethernet header */
typedef struct pcap2mjr_ethernet_header {
	uint8_t dst[6];
	uint8_t src[6];
	uint16_t type;
} pcap2mjr_ethernet_header;


/* Signal handler */
static void janus_p2m_handle_signal(int signum) {
	working = 0;
}

/* Main Code */
int main(int argc, char *argv[])
{
	struct gengetopt_args_info args_info;
	/* Let's call our cmdline parser */
	if(cmdline_parser(argc, argv, &args_info) != 0)
		exit(1);

	janus_log_init(FALSE, TRUE, NULL);
	atexit(janus_log_destroy);

	/* Evaluate arguments to find source and target */
	uint32_t ssrc = args_info.ssrc_arg;
	const char *codec = args_info.codec_arg;
	gboolean show_warnings = args_info.warnings_given;
	gboolean video = FALSE;
	if(!strcasecmp(codec, "vp8") || !strcasecmp(codec, "vp9") || !strcasecmp(codec, "h264")
			|| !strcasecmp(codec, "av1") || !strcasecmp(codec, "h265")) {
		video = TRUE;
	} else if(!strcasecmp(codec, "opus") || !strcasecmp(codec, "multiopus")
			|| !strcasecmp(codec, "g711") || !strcasecmp(codec, "pcmu") || !strcasecmp(codec, "pcma")
			|| !strcasecmp(codec, "g722")) {
		video = FALSE;
	} else if(!strcasecmp(codec, "text") || !strcasecmp(codec, "binary")) {
		/* We only do processing for RTP */
		JANUS_LOG(LOG_ERR, "Data channels not supported by this tool\n");
		cmdline_parser_free(&args_info);
		exit(1);
	} else {
		JANUS_LOG(LOG_ERR, "Unsupported codec '%s'\n", codec);
		cmdline_parser_free(&args_info);
		exit(1);
	}
	char *source = NULL, *destination = NULL, *setting = NULL;
	int i=0;
	for(i=1; i<argc; i++) {
		if(argv[i] == NULL || strlen(argv[i]) == 0) {
			setting = NULL;
			continue;
		}
		if(argv[i][0] == '-') {
			setting = argv[i];
			continue;
		}
		if(setting == NULL || (
				(strcmp(setting, "-c")) && (strcmp(setting, "--codec")) &&
				(strcmp(setting, "-s")) && (strcmp(setting, "--ssrc"))
		)) {
			if(source == NULL)
				source = argv[i];
			else if(destination == NULL)
				destination = argv[i];
		}
		setting = NULL;
	}
	if(source == NULL || destination == NULL) {
		cmdline_parser_print_help();
		cmdline_parser_free(&args_info);
		exit(1);
	}
	JANUS_LOG(LOG_INFO, "[%s/%"SCNu32"] %s --> %s\n", codec, ssrc, source, destination);

	/* Open and parse the pcap file */
	char errbuf[PCAP_ERRBUF_SIZE];
	pcap_t *pcap = pcap_open_offline(source, errbuf);
	if(pcap == NULL) {
		JANUS_LOG(LOG_ERR, "Could not open file %s: %s\n", source, errbuf);
		cmdline_parser_free(&args_info);
		exit(1);
	}
	int link = pcap_datalink(pcap);
	if(link != DLT_LINUX_SLL && link != DLT_EN10MB) {
		JANUS_LOG(LOG_ERR, "Unsupported link type %d (%s) in capture\n",
			link, pcap_datalink_val_to_name(link));
		cmdline_parser_free(&args_info);
		exit(1);
	}

	/* Create the target file */
	FILE *outfile = fopen(destination, "wb");
	if(outfile == NULL) {
		JANUS_LOG(LOG_ERR, "Couldn't open output file\n");
		cmdline_parser_free(&args_info);
		pcap_close(pcap);
		exit(1);
	}
	/* Write the first part of the header */
	size_t res = fwrite(header, sizeof(char), strlen(header), outfile);
	if(res != strlen(header)) {
		JANUS_LOG(LOG_ERR, "Couldn't write .mjr header (%zu != %zu, %s)\n",
			res, strlen(header), g_strerror(errno));
		cmdline_parser_free(&args_info);
		pcap_close(pcap);
		exit(1);
	}

	/* Handle SIGINT */
	working = 1;
	signal(SIGINT, janus_p2m_handle_signal);

	/* TODO Loop */
    struct pcap_pkthdr *header = NULL;
    const u_char *buffer = NULL, *temp = NULL;
	uint32_t count = 0, written = 0, pssrc = 0;
    int ret = 0;
    size_t min_size = sizeof(pcap2mjr_ethernet_header) + sizeof(struct iphdr) +
		sizeof(struct udphdr) + 12, pkt_size = 0;
	gboolean header_written = FALSE;
	gint64 start_ts = 0, pkt_ts = 0;
    while(working && (ret = pcap_next_ex(pcap, &header, &buffer)) >= 0) {
		count++;
		if(header->len != header->caplen) {
			if(show_warnings) {
				JANUS_LOG(LOG_WARN, "Packet and capture lengths differ (%d != %d), skipping packet #%"SCNu32"\n",
					header->len, header->caplen, count);
			}
			continue;
		}
		if(header->len < min_size) {
			if(show_warnings) {
				JANUS_LOG(LOG_WARN, "Packet too small (< %zu), skipping packet #%"SCNu32"\n", min_size, count);
			}
			continue;
		}
		temp = buffer;
		pkt_size = header->len;
		pkt_ts = header->ts.tv_sec*G_USEC_PER_SEC + header->ts.tv_usec;
		if(start_ts == 0)
			start_ts = pkt_ts;
		/* Traverse all the headers */
		int protocol = 0;
		if(link == DLT_EN10MB) {
			/* Ethernet */
			pcap2mjr_ethernet_header *eth = (pcap2mjr_ethernet_header *)temp;
			protocol = ntohs(eth->type);
			temp += sizeof(pcap2mjr_ethernet_header);
			pkt_size -= sizeof(pcap2mjr_ethernet_header);
		} else {
			/* Linux Cooked Capture */
			struct sll_header *lcc = (struct sll_header *)temp;
			protocol = ntohs(lcc->sll_protocol);
			temp += sizeof(struct sll_header);
			pkt_size -= sizeof(struct sll_header);
		}
		if(protocol == 0x0800) {
			/* IPv4 */
			struct iphdr *v4 = (struct iphdr *)temp;
			protocol = v4->protocol;
			temp += sizeof(struct iphdr);
			pkt_size -= sizeof(struct iphdr);
		} else if(protocol == 0x86DD) {
			/* IPv6 */
			struct ip6_hdr *v6 = (struct ip6_hdr *)temp;
			protocol = v6->ip6_ctlun.ip6_un1.ip6_un1_nxt;
			temp += sizeof(struct ip6_hdr);
			pkt_size -= sizeof(struct ip6_hdr);
		} else {
			if(show_warnings) {
				JANUS_LOG(LOG_WARN, "Not an IPv4 or IPv6 packet, skipping\n");
			}
			continue;
		}
		if(protocol != 17) {
			if(show_warnings) {
				JANUS_LOG(LOG_WARN, "Not an UDP packet, skipping\n");
			}
			continue;
		}
		/* UDP */
		temp += sizeof(struct udphdr);
		pkt_size -= sizeof(struct udphdr);
		/* Make sure this is an RTP packet */
		janus_pp_rtp_header *rtp = (janus_pp_rtp_header *)temp;
		if(rtp->version != 2 || (rtp->type >= 64 && rtp->type < 96)) {
			if(show_warnings) {
				JANUS_LOG(LOG_WARN, "Not an RTP packet, skipping packet #%"SCNu32"\n", count);
			}
			continue;
		}
		pssrc = htonl(rtp->ssrc);
		if(pssrc != ssrc) {
			if(show_warnings) {
				JANUS_LOG(LOG_WARN, "Not the SSRC we need (%"SCNu32" != %"SCNu32"), skipping packet #%"SCNu32"\n",
					pssrc, ssrc, count);
			}
			continue;
		}
		/* Save the packet, but first check if we've written the .mjr header already */
		if(!header_written) {
			/* Write info header as a JSON formatted info */
			header_written = TRUE;
			json_t *info = json_object();
			/* FIXME Codecs should be configurable in the future */
			const char *type = NULL;
			if(video)
				type = "v";
			else
				type = "a";
			json_object_set_new(info, "t", json_string(type));
			json_object_set_new(info, "c", json_string(codec));
			json_object_set_new(info, "s", json_integer(pkt_ts));
			json_object_set_new(info, "u", json_integer(pkt_ts));
			gchar *info_text = json_dumps(info, JSON_PRESERVE_ORDER);
			json_decref(info);
			uint16_t info_bytes = htons(strlen(info_text));
			size_t res = fwrite(&info_bytes, sizeof(uint16_t), 1, outfile);
			if(res != 1) {
				JANUS_LOG(LOG_WARN, "Couldn't write size of JSON header in .mjr file (%zu != %zu, %s), expect issues post-processing\n",
					res, sizeof(uint16_t), g_strerror(errno));
			}
			res = fwrite(info_text, sizeof(char), strlen(info_text), outfile);
			if(res != strlen(info_text)) {
				JANUS_LOG(LOG_WARN, "Couldn't write JSON header in .mjr file (%zu != %zu, %s), expect issues post-processing\n",
					res, strlen(info_text), g_strerror(errno));
			}
			free(info_text);
		}
		/* Write frame header (fixed part[4], timestamp[4], length[2]) */
		size_t res = fwrite(frame_header, sizeof(char), strlen(frame_header), outfile);
		if(res != strlen(frame_header)) {
			JANUS_LOG(LOG_WARN, "Couldn't write frame header in .mjr file (%zu != %zu, %s), expect issues post-processing\n",
				res, strlen(frame_header), g_strerror(errno));
		}
		uint32_t timestamp = (uint32_t)(pkt_ts > start_ts ? ((pkt_ts - start_ts)/1000) : 0);
		timestamp = htonl(timestamp);
		res = fwrite(&timestamp, sizeof(uint32_t), 1, outfile);
		if(res != 1) {
			JANUS_LOG(LOG_WARN, "Couldn't write frame timestamp in .mjr file (%zu != %zu, %s), expect issues post-processing\n",
				res, sizeof(uint32_t), g_strerror(errno));
		}
		uint16_t header_bytes = htons(pkt_size);
		res = fwrite(&header_bytes, sizeof(uint16_t), 1, outfile);
		if(res != 1) {
			JANUS_LOG(LOG_WARN, "Couldn't write size of frame in .mjr file (%zu != %zu, %s), expect issues post-processing\n",
				res, sizeof(uint16_t), g_strerror(errno));
		}
		/* Save packet on file */
		written++;
		int tmp = 0, tot = pkt_size;
		while(tot > 0) {
			tmp = fwrite(temp+pkt_size-tot, sizeof(char), tot, outfile);
			if(tmp <= 0) {
				JANUS_LOG(LOG_ERR, "Error saving frame, stopping here...\n");
				goto done;
			}
			tot -= tmp;
		}
	}
	JANUS_LOG(LOG_INFO, "Saved %"SCNu32" out of %"SCNu32" packets\n", written, count);

done:
	/* We're done */
	pcap_close(pcap);
	fclose(outfile);
	outfile = fopen(destination, "rb");
	if(outfile == NULL) {
		JANUS_LOG(LOG_WARN, "No destination file %s??\n", destination);
	} else {
		fseek(outfile, 0L, SEEK_END);
		size_t fsize = ftell(outfile);
		fseek(outfile, 0L, SEEK_SET);
		JANUS_LOG(LOG_INFO, "%s is %zu bytes\n", destination, fsize);
		fclose(outfile);
	}

	cmdline_parser_free(&args_info);
	JANUS_LOG(LOG_INFO, "Bye!\n");
	return 0;
}
